### PR TITLE
feat(orderbook): add sanity checks to placeOrder

### PR DESCRIPTION
**Pull request by ghost**
In this PR we're going to use the `maximumOutboundCapacity` from BaseClient to make sure that when placing the order sufficient balance exists. https://github.com/ExchangeUnion/xud/issues/675

There's 1 test case currently failing within [Service.spec](https://github.com/ExchangeUnion/xud/blob/master/test/integration/Service.spec.ts#L73) because the code now assumes reachable and funded swap client, but the test setup does not enable them. I'm open to suggestions on how to move forward with modifications within Service.spec. Should we enable lnds within test setup config and assume channels exist and are funded?

### CODE REVIEW

**Comment by kilrau** [comment]
Missing `maximumInboundCapacity`?

File: lib/orderbook/OrderBook.ts [hunk 1]
```diff
@@ -298,6 +298,17 @@ class OrderBook extends EventEmitter {
       };
     }
 
+    // check if sufficient outbound channel capacity exists
+    const { makerAmount } = Swaps.calculateSwapAmounts(order.quantity, order.price, order.isBuy);
+    const { makerCurrency } = Swaps.deriveCurrencies(order.pairId, order.isBuy);
+    const swapClient = this.swaps && this.swaps.swapClients[makerCurrency];
```
Comment (sangaman) [inline lib/orderbook/OrderBook.ts]: For getting the tests to pass, how about wrapping this block of code in an `if (this.swaps)` check and dropping it from this line?

**Review by sangaman — COMMENTED** [review]
Is the plan to put in similar checks before we add peer orders to the order book in a separate PR?

**Comment by erkarl** [comment]
> Missing maximumInboundCapacity?

Do we need to include the check for inbound capacity if the other peer is already also checking its outbound (our inbound) capacity?

**Comment by sangaman** [comment]
> Does it make sense include the check for inbound capacity if the other peer is already also checking its outbound (our inbound) capacity?

I'd be ok with leaving it up to the peer to check their side of the available capacity, particularly because LND doesn't give us a simple way of fetching the inbound capactiy (we'd have to list all channels and calculate it ourselves).

**Comment by kilrau** [comment]
> Does it make sense include the check for inbound capacity if the other peer is already also checking its outbound (our inbound) capacity?

Generally speaking: `don't trust, verify` , but

> LND doesn't give us a simple way of fetching the inbound capactiy (we'd have to list all channels and calculate it ourselves).

doing that regularly is a bit over the top, agree.

**Comment by kilrau** [comment]
> Is the plan to put in similar checks before we add peer orders to the order book in a separate PR?

Well here's the thing: `ChannelBalance` only matters in the moment of a match, when an actual swap is triggered. And a match only occurs upon `OwnOrder` entering the order book, never when a `PeerOrder` enters the order book. So letting an "out of channel bounds" `PeerOrder` enter our local order book is totally fine, checks happen when `OwnOrder` comes in and if "within channel bounds" it can still fill this over-sized `PeerOrder` partially.

Unless I'm missing something, I believe it's the right thing to NOT do this check on `PeerOrders`.

File: lib/orderbook/OrderBook.ts (hunk 1 above)
Comment (erkarl) [reply lib/orderbook/OrderBook.ts]: This won't work because the service spec is firing up a real instance of xud (with config defined in the test setup). It will have `this.swaps`.

Comment (erkarl) [reply lib/orderbook/OrderBook.ts]: I've updated this PR/commit to make sanity checks configurable from xud.conf (`sanitychecks = true`). This allows us to disable those checks in the service spec.

Alternative option is to remove the offending tests from service spec and add them to xud-simulation... 🤔

**Comment by kilrau** [comment]
[Above](https://github.com/ExchangeUnion/xud/pull/863#issuecomment-479892229) I said channel capacity checks upon order entering shouldn't be done on incoming peer orders, just on `OwnOrder`s since incoming matching `OwnOrder`s directly trigger a swap and only my `OwnOrder` size matters for the channel capacity - I can fill oversized `PeerOrder`s partially. Does that make sense? @erkarl 

Also as @erkarl pointed out before, the `getRoutes()` check on maker side when accepting a deal covers the capacity check back to the taker. So I think we don't have to do another `getRoutes()` check from taker to maker. The sanity check cover it. 

Bottom line I think we can safely replace [this in-code comment](https://github.com/ExchangeUnion/xud/blob/master/lib/swaps/Swaps.ts#L387) in this PR with an explainer that the sanity checks already cover the local amount of the channel.

**Comment by erkarl** [comment]
@kilrau 
In general I agree with the logic above, but having outbound balance does not always mean that routes exist (assuming non-direct channels). Therefore, I'd leave the TODO in place as a future improvement to be made.

@sangaman - rebased and ready for review again.

**Comment by sangaman** [comment]
I think that TODO comment can be removed with #822, as we'll be verifying that a route exists before accepting *any* orders from a peer as part of the sanity swap.

File: lib/Config.ts [hunk 2]
```diff
@@ -27,6 +27,8 @@ class Config {
   public dbpath: string;
   /** Whether matching will be disabled */
   public nomatching: boolean;
+  /** Whether to perform sanity checks */
+  public sanitychecks: boolean;
```
Comment (sangaman) [inline lib/Config.ts]: "sanity checks" is a little ambiguous but maybe I'll use this same option to disable the sanity swaps in which case it's a good thing it's not too specific. Maybe just add a some detail to the comment like "Whether to check that orders can possibly be swapped before adding them to the order book, can be disabled for testing & debugging purposes."

Another consideration is that if we ever pass a value for this option from the command line (which I think could happen if I reuse it for disabling #822) then we might want to call it `nosanitychecks` so that it can be set with `--nosanitychecks` from the command line.

**Review by sangaman — COMMENTED** [review]
Looks good but can you look into why the simulation tests are failing? It's not immediately obvious to me, but it's failing due to insufficient balance.

File: lib/Config.ts (hunk 2 above)
Comment (erkarl) [reply lib/Config.ts]: Good idea. I've renamed the config option to `nosanitychecks` and updated the comment.

**Comment by erkarl** [comment]
@sangaman - build is passing now.

**Review by sangaman — APPROVED** [review]
Awesome, thank you.
